store/etcd/put.go

```diff
@@ -3,236 +3,156 @@ package etcd
+	var err error
 
-		err = s.putEntry(stm, &re)
-		if err != nil {
-			return err
-		}
+	entry := store.RulesetEntry{
+		Path:    path,
+		Ruleset: ruleset,
+	}
 
-		// update the pointer to the latest ruleset
-		stm.Put(s.latestRulesetPath(path), s.entriesPath(path, version))
+	// validate the ruleset
+	entry.Signature, err = p.validateRuleset(p.stm, path, ruleset)
+	if err != nil {
+		return nil, err
+	}
 
-		entry = re
-		return nil
+	// encode rules
+	data, err := proto.Marshal(rulesetToProtobuf(ruleset))
+	if err != nil {
+		return nil, err
 	}
 
-	_, err = concurrency.NewSTM(s.Client, txfn, concurrency.WithAbortContext(ctx))
-	if err != nil && err != store.ErrNotModified && !store.IsValidationError(err) {
-		return nil, errors.Wrap(err, "failed to put ruleset")
+	// update checksum if rules have changed
+	changed, err := p.updateChecksum(p.stm, path, data)
+	if err != nil {
+		return nil, err
 	}
 
-	return &entry, err
-}
+	if !changed {
+		// fetch latest version string
+		_, entry.Version = p.s.pathVersionFromKey((p.stm.Get(p.s.latestVersionPath(path))))
 
-func compareSignature(base, other *regula.Signature) error {
-	if base.ReturnType != other.ReturnType {
-		return &store.ValidationError{
-			Field:  "return type",
-			Value:  other.ReturnType,
-			Reason: fmt.Sprintf("signature mismatch: return type must be of type %s", base.ReturnType),
-		}
+		return &entry, store.ErrRulesetNotModified
 	}
 
-	for name, tp := range other.ParamTypes {
-		stp, ok := base.ParamTypes[name]
-		if !ok {
-			return &store.ValidationError{
-				Field:  "param",
-				Value:  name,
-				Reason: "signature mismatch: unknown parameter",
-			}
-		}
-
-		if tp != stp {
-			return &store.ValidationError{
-				Field:  "param type",
-				Value:  tp,
-				Reason: fmt.Sprintf("signature mismatch: param must be of type %s", stp),
-			}
-		}
+	// create a new version of the ruleset
+	entry.Version, err = p.createNewVersion(p.stm, path, data)
+	if err != nil {
+		return nil, err
 	}
 
-	return nil
+	// update the pointer to the latest ruleset version
+	p.stm.Put(p.s.latestVersionPath(path), p.s.rulesetsPath(path, entry.Version))
+
+	return &entry, p.updateVersionRegistry(p.stm, path, entry.Version)
 }
 
-func validateRuleset(path string, rs *regula.Ruleset) (*regula.Signature, error) {
-	err := validateRulesetName(path)
-	if err != nil {
-		return nil, err
+// validateRules fetches the signature from the store and validates all the rules against it.
+// if the rules are valid, it returns the signature.
+func (p *rulesPutter) validateRuleset(stm concurrency.STM, path string, ruleset *regula.Ruleset) (*regula.Signature, error) {
+	data := stm.Get(p.s.signaturesPath(path))
+	if data == "" {
+		return nil, store.ErrSignatureNotFound
 	}
 
-	sig := regula.NewSignature(rs)
+	var pbsig pb.Signature
+	err := proto.Unmarshal([]byte(data), &pbsig)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to decode signature")
```

Comment:
Usually we use `unmarshall` instead of `decode`. 
I am not saying that this one is better than the other but we should agree on which one we use.

Reply:
Unmarshal is just more specific than decode, what we are actually doing is decoding, it is a more common word than unmarshal.